### DISCUSSION TITLE
The TOTP key in the csv file of keepassxc is not imported

### DISCUSSION THREAD

**Discussion by wc7086** [dani-garcia/vaultwarden#1941] [Q&A] [answered]
### Subject of the issue

The TOTP key in the csv file of keepassxc is not imported

### Deployment environment
* Vaultwarden version: v1.22.2
* Web-vault version: v2.21.1
* Running within Docker: true
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://**.****.***",
  "domain_origin": "*****://**.****.***",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": null,
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce

```bash
docker run -d --name vaultwarden -e WEBSOCKET_ENABLED=true -e DOMAIN=https://vw.example.com -e SIGNUPS_ALLOWED=false -e ADMIN_TOKEN=****** -v /opt/vw-data/:/data/ -p 8081:80 -p 3012:3012 vaultwarden/server:latest
```

### Expected behaviour

The TOTP key in the csv file of keepassxc can be imported

### Actual behaviour

The TOTP key in the csv file of keepassxc is not imported

**Comment by BlackDex** [comment, marked as answer]
That is not a Vaultwarden issue. The web-vault handels the import encrypt it and send it to the backend server.

As far as i know no extra keys and attachments are imported from any type of export.

**Reply by brianjmurrell** [reply]
I'm not sure I understand the answer here.  How is the **import** feature/function in a Vaultwarden installation not importing the TOTP field from a KeePassXC CSV export file not a Vaultwarden issue?  Surely it's Vaultwarden that is doing this import is it not?  Isn't it Vaultwarden that is reading the CSV file and putting each of the items (columns) from the CSV file into the Vaultwarden vault, line-by-line?

**Reply by brianjmurrell** [reply]
Oh.  Is the distinction here Vaultwarden-web vs. Vaultwarden (proper)?  Are you saying that this is a Vaultwarden-web issue?

**Reply by BlackDex** [reply]
It is a web-vault issue. The web-vault is not developed by this project. We only make some small patches so it will work with Vaultwarden.

So, if anything else like import or export isn't working that well it mostly is on the client side, since the clients send basic Bitwarden entries, which all work fine.

The conversion is done client side.

**Comment by florian-bellencontre** [comment]
Hello,

It seems that I have the same problem. Could you tell us what format is accepted to import a TOTP in CSV?

My TOTP looks like this in my KeepassXC export:
```
otpauth://totp/website.example:USERNAME?secret=SECRET&period=30&digits=6&issuer=website.example
```

**Comment by BlackDex** [comment]
Bitwarden doesn't support these **Custom** fields as a TOTP field.
See: https://community.bitwarden.com/t/smart-import-from-keepass/9289/3

**Comment by florian-bellencontre** [comment]
Do you know a simple way to convert these custom fileds to TOTP?

Or maybe you know where I can ask the question? On the bitwarden forum?

**Comment by ralphocdol** [comment]
I'm not sure if this is fully answered, but for those who came across this, you can try what I did, which is export it to XML.

I can't seem to export it as an XML using KeePassXC, so I tried looking for a different client and ended up with "KeePass Password Safe".
![KeePass](https://user-images.githubusercontent.com/31056871/198892990-0294e3cc-2e9d-46ad-9c61-c447976451c0.png)
